tests/transforms/test_split_non_commuting.py

```diff
@@ -0,0 +1,241 @@
+    qml.PauliZ(0)
+    qml.Hadamard(0)
+    qml.CNOT((0, 1))
+    qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+    qml.expval(qml.PauliX(0) @ qml.PauliX(1))
+    qml.expval(qml.PauliY(0) @ qml.PauliY(1))
+    qml.expval(qml.PauliZ(0))
+    qml.expval(qml.PauliX(0))
+    qml.expval(qml.PauliY(0))
+
+### example tape with 2 -commuting groups [[0,2],[1,3]]
+with qml.tape.QuantumTape() as non_commuting_tape2:
+    qml.PauliZ(0)
+    qml.Hadamard(0)
+    qml.CNOT((0, 1))
+    qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+    qml.expval(qml.PauliX(0) @ qml.PauliX(1))
+    qml.expval(qml.PauliZ(0))
+    qml.expval(qml.PauliX(0))
+
+non_commuting_tapes = [non_commuting_tape2, non_commuting_tape3]
+
+
+@pytest.mark.parametrize("tape,expected", [(non_commuting_tape2, 2), (non_commuting_tape3, 3)])
+def test_non_commuting_group_right_number(tape, expected):
+    """Test that the output is of the correct size"""
+    split, _ = split_non_commuting(tape)
+    assert len(split) == expected
+
+
+@pytest.mark.parametrize(
+    "tape,group_coeffs",
+    [(non_commuting_tape2, [[0, 2], [1, 3]]), (non_commuting_tape3, [[0, 3], [1, 4], [2, 5]])],
+)
+def test_non_commuting_group_right_reorder(tape, group_coeffs):
+    """Test that the output is of the correct order"""
+    split, fn = split_non_commuting(tape)
+    assert all(np.array(fn(group_coeffs)) == np.arange(len(split) * 2))
+
+
+### Test for other measurement types
+
+obs_fn = [qml.expval, qml.var, qml.sample]
+
+
+@pytest.mark.parametrize("meas_type", obs_fn)
+def test_different_measurement_types(meas_type):
+    """Test that expval, var and sample are correctly reproduced"""
+    with qml.tape.QuantumTape() as tape:
+        qml.PauliZ(0)
+        qml.Hadamard(0)
+        qml.CNOT((0, 1))
+        meas_type(qml.PauliZ(0) @ qml.PauliZ(1))
+        meas_type(qml.PauliX(0) @ qml.PauliX(1))
+        meas_type(qml.PauliZ(0))
+        meas_type(qml.PauliX(0))
+    the_return_type = tape.measurements[0].return_type
+    split, _ = split_non_commuting(tape)
+    for new_tape in split:
+        for meas in new_tape.measurements:
+            assert meas.return_type == the_return_type
+
+
+## Testing in context of qnode with groups of non-commuting observables
+
+
+def test_expval_non_commuting_observables():
+    """Test expval with multiple non-commuting operators"""
+    dev = qml.device("default.qubit", wires=6)
+
+    @qml.qnode(dev)
+    def circuit():
+        qml.Hadamard(1)
+        qml.Hadamard(0)
+        qml.PauliZ(0)
+        qml.Hadamard(3)
+        qml.Hadamard(5)
+        qml.T(5)
+        return [
+            qml.expval(qml.PauliZ(0) @ qml.PauliZ(1)),
+            qml.expval(qml.PauliX(0)),
+            qml.expval(qml.PauliZ(1)),
+            qml.expval(qml.PauliX(1) @ qml.PauliX(4)),
+            qml.expval(qml.PauliX(3)),
+            qml.expval(qml.PauliY(5)),
+        ]
+
+    assert all(np.isclose(circuit(), np.array([0.0, -1.0, 0.0, 0.0, 1.0, 1 / np.sqrt(2)])))
```

Comment:
@Qottmann how did you compute this statevector?

Reply:
There are no entangling gates, so this can be done in the head since all qubits are independent: 
The Pauli-Z expvals are 0 because I put a Hadamard on wire 0 and 1.  
The PauliX(0) is -1 because I put a Hadamard and Z on wire 0  
The PauliX(4) is 0 because nothing is done on wire 4  
The PauliY(5) is 1/sqrt(2) since I put a Hadamard and T gate on it.

The funny business with the tensor products is to create rich, non-trivial groupings.

Reply:
nice! a very good way of writing the test (checking against an analytic quantity). I just wanted to ensure this wasn't the result of running PL with `qml.state()` lol